File: sources/Diablo 2.1/net/minecraft/client/gui/GuiMainMenu.java
package net.minecraft.client.gui;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldRenderer;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.MathHelper;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL20;
import org.lwjgl.util.glu.Project;
import wtf.diablo.Diablo;
import wtf.diablo.gui.GuiAltManager;
import wtf.diablo.gui.mainMenu.MainMenuButton;
import wtf.diablo.utils.font.Fonts;
import wtf.diablo.utils.glsl.GLSLSandboxShader;
import wtf.diablo.utils.render.ChatColor;
import wtf.diablo.utils.render.ColorUtil;
import wtf.diablo.utils.render.RenderUtil;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Random;

public class GuiMainMenu extends GuiScreen implements GuiYesNoCallback
{
    private long initTime = System.currentTimeMillis();
    private GLSLSandboxShader backgroundShader;
    public ArrayList<MainMenuButton> buttons = new ArrayList<MainMenuButton>();
    private boolean supportsGL = true;
    private static final ResourceLocation minecraftTitleTextures = new ResourceLocation("textures/gui/title/minecraft.png");
    private ResourceLocation backgroundTexture;
    private DynamicTexture viewportTexture;
    private int panoramaTimer;
    private float updateCounter;
    private static final Random RANDOM = new Random();

    private static final ResourceLocation[] titlePanoramaPaths = new ResourceLocation[] {new ResourceLocation("textures/gui/title/background/panorama_0.png"), new ResourceLocation("textures/gui/title/background/panorama_1.png"), new ResourceLocation("textures/gui/title/background/panorama_2.png"), new ResourceLocation("textures/gui/title/background/panorama_3.png"), new ResourceLocation("textures/gui/title/background/panorama_4.png"), new ResourceLocation("textures/gui/title/background/panorama_5.png")};

    public GuiMainMenu()
    {
        try {
            this.backgroundShader = new GLSLSandboxShader("/assets/minecraft/diablo/shader/background.fsh");
        } catch (Exception e) {
            e.printStackTrace();
        }
        this.updateCounter = RANDOM.nextFloat();
    }



    public void initGui() {
        this.viewportTexture = new DynamicTexture(256, 256);
        this.backgroundTexture = mc.getTextureManager().getDynamicTextureLocation("background", this.viewportTexture);
        buttons.clear();
        addButtons(height - 60);

        initTime = System.currentTimeMillis();
    }

    private void addButtons(int y) {
        double number = (width /5f) - 8;
        buttons.add(new MainMenuButton(0,"Singleplayer", 10, y, number,40));
        buttons.add(new MainMenuButton(1,"Multiplayer", 15 + number, y, number,40));
        buttons.add(new MainMenuButton(2,"Alt Manager", 20 + number * 2, y, number,40));
        buttons.add(new MainMenuButton(3,"Settings", 25 + number * 3, y, number,40));
        buttons.add(new MainMenuButton(4,"Exit", 30 + number * 4, y, number,40));
    }


    /**
     * Called by the controls from the buttonList when activated. (Mouse pressed for buttons)
     */
    protected void actionPerformed(int buttonID) throws IOException {
        switch (buttonID) {
            case 0:
                mc.displayGuiScreen(new GuiSelectWorld(this));
                break;
            case 1:
                mc.displayGuiScreen(new GuiMultiplayer(this));
                break;
            case 2:
                mc.displayGuiScreen(new GuiAltManager(this));
                break;
            case 3:
                mc.displayGuiScreen(new GuiOptions(this, mc.gameSettings));
                break;
            case 4:
                mc.shutdown();
                break;
        }
    }


    public void confirmClicked(boolean result, int id) {
    }
    private void rotateAndBlurSkybox(float p_73968_1_)
    {
        this.mc.getTextureManager().bindTexture(this.backgroundTexture);
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_LINEAR);
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR);
        GL11.glCopyTexSubImage2D(GL11.GL_TEXTURE_2D, 0, 0, 0, 0, 0, 256, 256);
        GlStateManager.enableBlend();
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);
        GlStateManager.colorMask(true, true, true, false);
        Tessellator tessellator = Tessellator.getInstance();
        WorldRenderer worldrenderer = tessellator.getWorldRenderer();
        worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR);
        GlStateManager.disableAlpha();
        int i = 3;

        for (int j = 0; j < i; ++j)
        {
            float f = 1.0F / (float)(j + 1);
            int k = this.width;
            int l = this.height;
            float f1 = (float)(j - i / 2) / 256.0F;
            worldrenderer.pos((double)k, (double)l, (double)this.zLevel).tex((double)(0.0F + f1), 1.0D).color(1.0F, 1.0F, 1.0F, f).endVertex();
            worldrenderer.pos((double)k, 0.0D, (double)this.zLevel).tex((double)(1.0F + f1), 1.0D).color(1.0F, 1.0F, 1.0F, f).endVertex();
            worldrenderer.pos(0.0D, 0.0D, (double)this.zLevel).tex((double)(1.0F + f1), 0.0D).color(1.0F, 1.0F, 1.0F, f).endVertex();
            worldrenderer.pos(0.0D, (double)l, (double)this.zLevel).tex((double)(0.0F + f1), 0.0D).color(1.0F, 1.0F, 1.0F, f).endVertex();
        }

        tessellator.draw();
        GlStateManager.enableAlpha();
        GlStateManager.colorMask(true, true, true, true);
    }
    private void drawPanorama(int p_73970_1_, int p_73970_2_, float p_73970_3_)
    {
        Tessellator tessellator = Tessellator.getInstance();
        WorldRenderer worldrenderer = tessellator.getWorldRenderer();
        GlStateManager.matrixMode(5889);
        GlStateManager.pushMatrix();
        GlStateManager.loadIdentity();
        Project.gluPerspective(120.0F, 1.0F, 0.05F, 10.0F);
        GlStateManager.matrixMode(5888);
        GlStateManager.pushMatrix();
        GlStateManager.loadIdentity();
        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
        GlStateManager.rotate(180.0F, 1.0F, 0.0F, 0.0F);
        GlStateManager.rotate(90.0F, 0.0F, 0.0F, 1.0F);
        GlStateManager.enableBlend();
        GlStateManager.disableAlpha();
        GlStateManager.disableCull();
        GlStateManager.depthMask(false);
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);
        int i = 8;

        for (int j = 0; j < i * i; ++j)
        {
            GlStateManager.pushMatrix();
            float f = ((float)(j % i) / (float)i - 0.5F) / 64.0F;
            float f1 = ((float)(j / i) / (float)i - 0.5F) / 64.0F;
            float f2 = 0.0F;
            GlStateManager.translate(f, f1, f2);
            GlStateManager.rotate(MathHelper.sin(((float)this.panoramaTimer + p_73970_3_) / 400.0F) * 25.0F + 20.0F, 1.0F, 0.0F, 0.0F);
            GlStateManager.rotate(-((float)this.panoramaTimer + p_73970_3_) * 0.1F, 0.0F, 1.0F, 0.0F);

            for (int k = 0; k < 6; ++k)
            {
                GlStateManager.pushMatrix();

                if (k == 1)
                {
                    GlStateManager.rotate(90.0F, 0.0F, 1.0F, 0.0F);
                }

                if (k == 2)
                {
                    GlStateManager.rotate(180.0F, 0.0F, 1.0F, 0.0F);
                }

                if (k == 3)
                {
                    GlStateManager.rotate(-90.0F, 0.0F, 1.0F, 0.0F);
                }

                if (k == 4)
                {
                    GlStateManager.rotate(90.0F, 1.0F, 0.0F, 0.0F);
                }

                if (k == 5)
                {
                    GlStateManager.rotate(-90.0F, 1.0F, 0.0F, 0.0F);
                }

                this.mc.getTextureManager().bindTexture(titlePanoramaPaths[k]);
                worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR);
                int l = 255 / (j + 1);
                float f3 = 0.0F;
                worldrenderer.pos(-1.0D, -1.0D, 1.0D).tex(0.0D, 0.0D).color(255, 255, 255, l).endVertex();
                worldrenderer.pos(1.0D, -1.0D, 1.0D).tex(1.0D, 0.0D).color(255, 255, 255, l).endVertex();
                worldrenderer.pos(1.0D, 1.0D, 1.0D).tex(1.0D, 1.0D).color(255, 255, 255, l).endVertex();
                worldrenderer.pos(-1.0D, 1.0D, 1.0D).tex(0.0D, 1.0D).color(255, 255, 255, l).endVertex();
                tessellator.draw();
                GlStateManager.popMatrix();
            }

            GlStateManager.popMatrix();
            GlStateManager.colorMask(true, true, true, false);
        }

        worldrenderer.setTranslation(0.0D, 0.0D, 0.0D);
        GlStateManager.colorMask(true, true, true, true);
        GlStateManager.matrixMode(5889);
        GlStateManager.popMatrix();
        GlStateManager.matrixMode(5888);
        GlStateManager.popMatrix();
        GlStateManager.depthMask(true);
        GlStateManager.enableCull();
        GlStateManager.enableDepth();
    }
    private void renderSkybox(int p_73971_1_, int p_73971_2_, float p_73971_3_)
    {
        this.mc.getFramebuffer().unbindFramebuffer();
        GlStateManager.viewport(0, 0, 256, 256);
        this.drawPanorama(p_73971_1_, p_73971_2_, p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.rotateAndBlurSkybox(p_73971_3_);
        this.mc.getFramebuffer().bindFramebuffer(true);
        GlStateManager.viewport(0, 0, this.mc.displayWidth, this.mc.displayHeight);
        float f = this.width > this.height ? 120.0F / (float)this.width : 120.0F / (float)this.height;
        float f1 = (float)this.height * f / 256.0F;
        float f2 = (float)this.width * f / 256.0F;
        int i = this.width;
        int j = this.height;
        Tessellator tessellator = Tessellator.getInstance();
        WorldRenderer worldrenderer = tessellator.getWorldRenderer();
        worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR);
        worldrenderer.pos(0.0D, (double)j, (double)this.zLevel).tex((double)(0.5F - f1), (double)(0.5F + f2)).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
        worldrenderer.pos((double)i, (double)j, (double)this.zLevel).tex((double)(0.5F - f1), (double)(0.5F - f2)).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
        worldrenderer.pos((double)i, 0.0D, (double)this.zLevel).tex((double)(0.5F + f1), (double)(0.5F - f2)).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
        worldrenderer.pos(0.0D, 0.0D, (double)this.zLevel).tex((double)(0.5F + f1), (double)(0.5F + f2)).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
        tessellator.draw();
    }




    /**
     * Draws the screen and all the components in it. Args : mouseX, mouseY, renderPartialTicks
     */
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        if (supportsGL)
            try {
                GlStateManager.enableAlpha();
                GlStateManager.disableCull();
                this.backgroundShader.useShader(width, height, mouseX, mouseY, (System.currentTimeMillis() - initTime) / 1000f);



                GL11.glBegin(GL11.GL_QUADS);

                GL11.glVertex2f(-1f, -1f);
                GL11.glVertex2f(-1f, 1f);
                GL11.glVertex2f(1f, 1f);
                GL11.glVertex2f(1f, -1f);

                GL11.glEnd();

                // Unbind shader
                GL20.glUseProgram(0);
            } catch (Exception ex) {
                ex.printStackTrace();

                GlStateManager.disableAlpha();
                this.renderSkybox(mouseX, mouseY, partialTicks);
                GlStateManager.enableAlpha();
                Tessellator tessellator = Tessellator.getInstance();
                WorldRenderer worldrenderer = tessellator.getWorldRenderer();
                int i = 274;
                int j = this.width / 2 - i / 2;
                int k = 30;
                this.drawGradientRect(0, 0, this.width, this.height, -2130706433, 16777215);
                this.drawGradientRect(0, 0, this.width, this.height, 0, Integer.MIN_VALUE);
                this.mc.getTextureManager().bindTexture(minecraftTitleTextures);
                GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);

                if ((double)this.updateCounter < 1.0E-4D)
                {
                    this.drawTexturedModalRect(j + 0, k + 0, 0, 0, 99, 44);
                    this.drawTexturedModalRect(j + 99, k + 0, 129, 0, 27, 44);
                    this.drawTexturedModalRect(j + 99 + 26, k + 0, 126, 0, 3, 44);
                    this.drawTexturedModalRect(j + 99 + 26 + 3, k + 0, 99, 0, 26, 44);
                    this.drawTexturedModalRect(j + 155, k + 0, 0, 45, 155, 44);
                }
                else
                {
                    this.drawTexturedModalRect(j + 0, k + 0, 0, 0, 155, 44);
                    this.drawTexturedModalRect(j + 155, k + 0, 0, 45, 155, 44);
                }

                GlStateManager.pushMatrix();
                GlStateManager.translate((float)(this.width / 2 + 90), 70.0F, 0.0F);
                GlStateManager.rotate(-20.0F, 0.0F, 0.0F, 1.0F);
                this.supportsGL = false;
            }
        //Fonts.SFReg45.drawStringWithShadow(Diablo.name + " " + Diablo.version, width / 2f - Fonts.SFReg45.getStringWidth(Diablo.name + " " + Diablo.version)/2f, height / 6f, 0xfFffffff);
        int imageSize = 100;
        if(height > 300)
            imageSize = 150;
        RenderUtil.drawImage(width / 2f - (imageSize / 2f), (height - 60) / 2f - (imageSize / 2f),imageSize,imageSize,new ResourceLocation("diablo/icons/logo.png"),1);

        String loggedInStr = ChatColor.WHITE + "Logged in as: " + ChatColor.GREEN + "Eclipse Ready by LuMi" + ChatColor.GRAY + " (" + "6969" + ChatColor.GRAY +")";

        mc.fontRendererObj.drawString(loggedInStr, (int) (width / 2f - (mc.fontRendererObj.getStringWidth(loggedInStr) / 2f)),height - 4 - mc.fontRendererObj.FONT_HEIGHT,-1);
        for(MainMenuButton but : buttons){
            but.drawScreen(mouseX,mouseY, this.width > 500);
        }
        super.drawScreen(mouseX, mouseY, partialTicks);
    }
    public void updateScreen()
    {
        ++this.panoramaTimer;
    }

    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        if(mouseButton == 0)
            for(MainMenuButton but : buttons){
                if(but.mouseClicked(mouseX,mouseY)){
                    actionPerformed(but.id);
                }
            }
        super.mouseClicked(mouseX, mouseY, mouseButton);
    }
}

File: sources/LiquidSense Dev/net/ccbluex/liquidbounce/ui/client/GuiMainMenu.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.ui.client

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.ui.client.altmanager.GuiAltManager
import net.ccbluex.liquidbounce.ui.font.Fonts
import net.ccbluex.liquidbounce.utils.render.RenderUtils.drawArc
import net.ccbluex.liquidbounce.utils.render.RenderUtils.getAnimationState
import net.minecraft.client.gui.*
import net.minecraft.client.resources.I18n
import org.lwjgl.opengl.GL11
import java.awt.Color


class GuiMainMenu : GuiScreen(), GuiYesNoCallback {

    override fun initGui() {
        addbutton()
        super.initGui()
    }
    /*

    var animation = 0.0
    var zoom = 0.0
    var slide = 0
    var sinOffset = 0
    var panoramaTimer = 0

    override fun updateScreen() {
        ++panoramaTimer
    }

    private fun drawPanorama(p_73970_1_: Int, p_73970_2_: Int, p_73970_3_: Float) {
        val tessellator = Tessellator.getInstance()
        val worldrenderer = tessellator.worldRenderer
        GlStateManager.matrixMode(5889)
        GlStateManager.pushMatrix()
        GlStateManager.loadIdentity()
        Project.gluPerspective(120.0f, 1.0f, 0.05f, 10.0f)
        GlStateManager.matrixMode(5888)
        GlStateManager.pushMatrix()
        GlStateManager.loadIdentity()
        GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
        GlStateManager.rotate(180.0f, 1.0f, 0.0f, 0.0f)
        GlStateManager.rotate(90.0f, 0.0f, 0.0f, 1.0f)
        GlStateManager.enableBlend()
        GlStateManager.disableAlpha()
        GlStateManager.disableCull()
        GlStateManager.depthMask(false)
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0)
        val i = 8
        for (j in 0 until i * i) {
            GlStateManager.pushMatrix()
            val f = ((j % i).toFloat() / i.toFloat() - 0.5f) / 64.0f
            val f1 = ((j / i)  / i.toFloat() - 0.5f) / 64.0f
            val f2 = 0.0f
            GlStateManager.translate(f, f1, f2)
            GlStateManager.rotate(MathHelper.sin((this.panoramaTimer + p_73970_3_) / 400.0f) * 25.0f + 20.0f,
                    1.0f, 0.0f, 0.0f)
            GlStateManager.rotate(-(this.panoramaTimer + p_73970_3_) * 0.1f, 0.0f, 1.0f, 0.0f)
            for (k in 0..5) {
                GlStateManager.pushMatrix()
                if (k == 1) {
                    GlStateManager.rotate(90.0f, 0.0f, 1.0f, 0.0f)
                }
                if (k == 2) {
                    GlStateManager.rotate(180.0f, 0.0f, 1.0f, 0.0f)
                }
                if (k == 3) {
                    GlStateManager.rotate(-90.0f, 0.0f, 1.0f, 0.0f)
                }
                if (k == 4) {
                    GlStateManager.rotate(90.0f, 1.0f, 0.0f, 0.0f)
                }
                if (k == 5) {
                    GlStateManager.rotate(-90.0f, 1.0f, 0.0f, 0.0f)
                }
                mc.textureManager.bindTexture(Gui.optionsBackground)
                worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR)
                val l = 255 / (j + 1)
                val f3 = 0.0f
                worldrenderer.pos(-1.0, -1.0, 1.0).tex(0.0, 0.0).color(255, 255, 255, l).endVertex()
                worldrenderer.pos(1.0, -1.0, 1.0).tex(1.0, 0.0).color(255, 255, 255, l).endVertex()
                worldrenderer.pos(1.0, 1.0, 1.0).tex(1.0, 1.0).color(255, 255, 255, l).endVertex()
                worldrenderer.pos(-1.0, 1.0, 1.0).tex(0.0, 1.0).color(255, 255, 255, l).endVertex()
                tessellator.draw()
                GlStateManager.popMatrix()
            }
            GlStateManager.popMatrix()
            GlStateManager.colorMask(true, true, true, false)
        }
        worldrenderer.setTranslation(0.0, 0.0, 0.0)
        GlStateManager.colorMask(true, true, true, true)
        GlStateManager.matrixMode(5889)
        GlStateManager.popMatrix()
        GlStateManager.matrixMode(5888)
        GlStateManager.popMatrix()
        GlStateManager.depthMask(true)
        GlStateManager.enableCull()
        GlStateManager.enableDepth()
    }

    private fun rotateAndBlurSkybox(p_73968_1_: Float) {
        mc.textureManager.bindTexture(Gui.optionsBackground)
        GL11.glTexParameteri(3553, 10241, 9729)
        GL11.glTexParameteri(3553, 10240, 9729)
        GL11.glCopyTexSubImage2D(3553, 0, 0, 0, 0, 0, 256, 256)
        GlStateManager.enableBlend()
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0)
        GlStateManager.colorMask(true, true, true, false)
        val tessellator = Tessellator.getInstance()
        val worldrenderer = tessellator.worldRenderer
        worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR)
        GlStateManager.disableAlpha()
        val i = 3
        for (j in 0 until i) {
            val f = 1.0f / (j + 1).toFloat()
            val k = width
            val l = height
            val f1 = (j - i / 2).toFloat() / 256.0f
            worldrenderer.pos(k.toDouble(), l.toDouble(), zLevel.toDouble()).tex(0.0f + f1.toDouble(), 1.0).color(1.0f, 1.0f, 1.0f, f).endVertex()
            worldrenderer.pos(k.toDouble(), 0.0, zLevel.toDouble()).tex(1.0f + f1.toDouble(), 1.0).color(1.0f, 1.0f, 1.0f, f).endVertex()
            worldrenderer.pos(0.0, 0.0, zLevel.toDouble()).tex(1.0f + f1.toDouble(), 0.0).color(1.0f, 1.0f, 1.0f, f).endVertex()
            worldrenderer.pos(0.0, l.toDouble(), zLevel.toDouble()).tex(0.0f + f1.toDouble(), 0.0).color(1.0f, 1.0f, 1.0f, f).endVertex()
        }
        tessellator.draw()
        GlStateManager.enableAlpha()
        GlStateManager.colorMask(true, true, true, true)
    }
    fun renderSkidBackground(mouseX: Int, mouseY: Int, partialTicks: Float) {
        this.mc.getFramebuffer().unbindFramebuffer();
        GlStateManager.viewport(0, 0, 256, 256);
        this.drawPanorama(mouseX, mouseY, partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.rotateAndBlurSkybox( partialTicks);
        this.mc.getFramebuffer().bindFramebuffer(true);
        if (this.animation < 4.0) {
            this.animation += 0.05000000074505806
        }
        this.zoom = if (this.zoom > 1.0) this.zoom * this.zoom * this.zoom * this.zoom / 350.0.let { this.zoom -= it; this.zoom } else 1.0
        val z = this.zoom
        val ScaledResolution = ScaledResolution(this.mc)
        Gui.drawScaledCustomSizeModalRect(0, 0,
                (-(ScaledResolution.getScaledWidth() / z / 2.0
                        + ScaledResolution.getScaledWidth() / 2.0)).toFloat(),
                (-(ScaledResolution.getScaledHeight() / z) / 2.0
                        + ScaledResolution.getScaledHeight() / 2.0).toFloat(),
                this.width, this.height, (this.width * z).toInt(), (this.height * z).toInt(), this.width.toFloat(),
                height.toFloat())
        val width: Int = ScaledResolution.getScaledWidth()
        var height: Int = ScaledResolution.getScaledHeight()
        height = ((height.toDouble() / (this.animation * this.animation + 1.2)).toInt())
        if (height < 20) {
            height = 20
        }

        var lines = 0
        GlStateManager.disableTexture2D()
        for (ball in Menu.instances) {
            ball.render()
            for (ball2 in Menu.instances) {
                if (lines >= 20 || ball.distance(mouseX.toDouble(), mouseY.toDouble()) >= 60.0
                        || ball2.distance(ball.getX(), ball.getY()) >= 50.0) continue
                ball.drawLineTo(ball2)
                ++lines
            }
        }
        GL11.glPushMatrix()
        GL11.glBegin(7)
        GL11.glColor4f(0.6f, 0.6f, 0.6f, 1.0f)
        GL11.glVertex2d(0.0, height.toDouble() + 0.5)
        GL11.glVertex2d(width.toDouble(), height.toDouble() + 0.5)
        GL11.glVertex2d(width.toDouble(), 0.0)
        GL11.glVertex2d(0.0, 0.0)
        GL11.glEnd()
        GL11.glBegin(7)
        //	GL11.glColor4f(r / 1.5f, g / 1.5f, b / 1.5f, 1.0f);
        GL11.glVertex2d(0.0, height.toDouble())
        GL11.glVertex2d(width.toDouble(), height.toDouble())
        GL11.glVertex2d(width.toDouble(), 0.0)
        GL11.glVertex2d(0.0, 0.0)
        GL11.glEnd()
        GL11.glBegin(7)
        GL11.glColor4f(0.6f, 0.6f, 0.6f, 1.0f)
        GL11.glVertex2d(0.0, ScaledResolution.getScaledHeight().toDouble())
        GL11.glVertex2d(width.toDouble(), ScaledResolution.getScaledHeight().toDouble())
        GL11.glVertex2d(width.toDouble(), (ScaledResolution.getScaledHeight() - height).toDouble() + 4.5)
        GL11.glVertex2d(0.0, (ScaledResolution.getScaledHeight() - height).toDouble() + 4.5)
        GL11.glEnd()
        GL11.glBegin(7)
        //GL11.glColor4f(r, g, b, 1.0f);
        GL11.glVertex2d(0.0, ScaledResolution.getScaledHeight().toDouble())
        GL11.glVertex2d(width.toDouble(), ScaledResolution.getScaledHeight().toDouble())
        GL11.glVertex2d(width.toDouble(), ScaledResolution.getScaledHeight() - height + 5.toDouble())
        GL11.glVertex2d(0.0, ScaledResolution.getScaledHeight() - height + 5.toDouble())
        GL11.glEnd()
        GlStateManager.enableTexture2D()
        GL11.glPopMatrix()
    }
    */


    fun addbutton(){
        this.buttonList.add(GuiButton(1, 0, 0, this.width / 8, 20, I18n.format("menu.singleplayer")))
        this.buttonList.add(GuiButton(2, (this.width / 2 - ((this.width / 8) / 2) - 0) / 2, 0, this.width / 8, 20, I18n.format("menu.multiplayer")))
        this.buttonList.add(GuiButton(100, this.width / 2 - ((this.width / 8) / 2) , 0, this.width / 8, 20, "AltManager"))
        this.buttonList.add(GuiButton(0, (this.width / 2 - ((this.width / 8) / 2)) + ((this.width / 2 - ((this.width / 8) / 2) - 0) / 2), 0,  this.width / 8, 20, I18n.format("menu.options")))
        this.buttonList.add(GuiButton(4, this.width - (this.width / 8), 0, this.width / 8, 20, I18n.format("menu.quit")))

    }

    var animation = 0.0
    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        //this.animation = getAnimationState(this.animation, 100.0, Math.max(10.0, Math.abs(this.animation - 100) * 30.0) * 0.3)
        drawBackground(0)
        /*
        renderSkidBackground(mouseX, mouseY,partialTicks)
         */
        //GL11.glPushMatrix()
        //GL11.glTranslatef((this.width / 2).toFloat(), (this.height / 2).toFloat(), 0.0f)
        //this.animation = RenderUtils.getAnimationState(this.animation, n.toDouble(), Math.max(10.0, Math.abs(this.animation - n) * 30.0) * 0.3)
        //drawArc(1.0f, 1.0f, (this.height / 5).toDouble(), Color.BLUE.rgb, 20, 180.0 + 3.5999999046325684 * animation, 10)
        //GL11.glPopMatrix()

        Gui.drawRect(0, 21, this.width, 22, Color(0, 0, 0).rgb)
        Gui.drawRect(0, this.height - 16, this.width, this.height - 15, Color(0, 0, 0).rgb)
        Gui.drawRect(0, 0, this.width, 22, Integer.MIN_VALUE)
        Gui.drawRect(0, this.height - 16, this.width, this.height, Integer.MIN_VALUE)
        Fonts.font35.drawCenteredString("DevelopMent By AquaVit", this.width / 2F, this.height - 20F + Fonts.font35.FONT_HEIGHT, 0xffffff, true)
        Fonts.fontBold180.drawCenteredString(LiquidBounce.CLIENT_NAME, this.width / 2F, height / 2F - 15, 4673984, true)
        Fonts.font35.drawCenteredString(LiquidBounce.CLIENT_VERSION, this.width / 2F + 148, height / 2F - 15 + Fonts.font35.FONT_HEIGHT, 0xffffff, true)
        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    override fun actionPerformed(button: GuiButton) {
        when (button.id) {
            0 -> mc.displayGuiScreen(GuiOptions(this, mc.gameSettings))
            1 -> mc.displayGuiScreen(GuiSelectWorld(this))
            2 -> mc.displayGuiScreen(GuiMultiplayer(this))
            4 -> mc.shutdown()
            100 -> mc.displayGuiScreen(GuiAltManager(this))
            101 -> mc.displayGuiScreen(GuiServerStatus(this))
            102 -> mc.displayGuiScreen(GuiBackground(this))
            103 -> mc.displayGuiScreen(GuiModsMenu(this))
            108 -> mc.displayGuiScreen(GuiCredits(this))
        }
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {}
}